chore: migrate from vue-cli to vite for building

Gavinok opened the pull request:
This PR resolves #241

with this we are now using vite and have some extra eslint settings to hopefully ease the migration to vue 3 process.

Gavinok committed:
Migrate from vue-cli to vite for building

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

esune reviewed (commented):
Changes look good, but I still have to run the project to test it.

esune commented:
Adding @loneil to the review as well since he has more experience with Vue 3

Gavinok commented:
Language translation doesn't seem to be working in some files so I am going to reset this back to the draft state for now while I figure that out

amanji commented:
When it's ready I'll try running locally

amanji commented on package.json:
```diff
@@ -21,44 +19,37 @@
-    "vue-template-compiler": "^2.7.16",
+    "unplugin-vue-components": "^0.27.4",
+    "vite": "^5.4.11",
+    "vue-i18n": "^8.28.2",
```
Should not be a dev dependency

amanji replied:
Referring to `vue-i18n` here

Gavinok replied:
I have migrated this from the dev deps. Thanks for noticing that

Gavinok committed:
Corrected vue-i18n from being a dev to a regular dependancy

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

loneil commented on package.json:
```diff
@@ -21,26 +19,21 @@
     "vue-router": "^3.6.5",
     "vuetify": "^2.7.2",
-    "vuex": "^3.6.2"
+    "vuex": "^3.6.2",
```
Would consider upgrade strategy for Vuex. version 3 hasn't been updated in 4 years (and version 4 in 2 years).
Pinia is now (see notes on Vuex docs https://vuex.vuejs.org/) the mainline state management.
So if the intent is to keep this app maintainable in terms of library updates for the future, Vuex might block that.

But of course switching to Pinia would involve additional redo work, but it might be comparable to the Vue 2->3 switch? Depends on appetite for rework vs old library.

Would at least switch to Vuex 4 if that's not a big lift, but really if the intent is to keep this up to date there's not much other than switching to Pinia from what I can see.

(just a note for future migration step when it's gotten to, not related to Vite update here)

Gavinok replied:
I had planned to migrate to pinia as part of the 3.0 migration but this seems like as good a time as any to add it.

Gavinok committed:
Fixed error where i18n translations kept failing

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

loneil commented on vue.config.js:
```diff
@@ -28,8 +28,8 @@ module.exports = {
-          // target: "https://orgbook.gov.bc.ca",
-          target: "http://host.docker.internal:8080",
+          target: "https://dev.orgbook.gov.bc.ca",
+          // target: "http://host.docker.internal:8080",
```
Intentional change swapping these for merging to main?

Gavinok replied:
I should just stick with the original local host option. My mistake

Gavinok committed:
Added pinia as a dependancy

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

Gavinok committed:
Removed old comments and default to localhost

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

Gavinok commented:
I have now resolved the issues with translation failing and this is now ready for review

loneil commented:
Don't really know this app so can't give too much of a test, but if I try a `manage build` I do get

```
16.83 x Build failed in 15.25s
16.83 error during build:
16.83 [vite:esbuild-transpile] Transform failed with 1 error:
16.83 assets/index-!~{001}~.js:19239:12: ERROR: Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides)     
16.83
16.83 Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides)
16.83 19237|    locale: "en",
16.83 19238|    fallbackLocale: "en",
16.83 19239|    messages: await loadLocaleMessages()
16.83    |              ^
16.83 19240|  });
16.83 19241|
16.83
16.83     at failureErrorWithLog (/app/node_modules/esbuild/lib/main.js:1472:15)
16.83     at /app/node_modules/esbuild/lib/main.js:755:50
16.83     at responseCallbacks.<computed> (/app/node_modules/esbuild/lib/main.js:622:9)
16.83     at handleIncomingPacket (/app/node_modules/esbuild/lib/main.js:677:12)
16.83     at Socket.readFromStdout (/app/node_modules/esbuild/lib/main.js:600:7)
16.83     at Socket.emit (node:events:513:28)
16.83     at addChunk (node:internal/streams/readable:315:12)
16.83     at readableAddChunk (node:internal/streams/readable:289:9)
16.83     at Socket.Readable.push (node:internal/streams/readable:228:10)
```

again, not sure if I'm doing things right (following readme instructions).

Also a ton of SASS related deprecation warnings get thrown but that's probably something to sort with the bulk code changes with library updates later.

@amanji cleaned some stuff up in Traction related to this in the past (I think, we did remove that loadLocaleMessages function to use the plugin), can see what he did here if it applies:
https://github.com/bcgov/traction/pull/630

amanji commented:
> Don't really know this app so can't give too much of a test, but if I try a `manage build` I do get
> 
> ```
> 16.83 x Build failed in 15.25s
> 16.83 error during build:
> 16.83 [vite:esbuild-transpile] Transform failed with 1 error:
> 16.83 assets/index-!~{001}~.js:19239:12: ERROR: Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides)     
> 16.83
> 16.83 Top-level await is not available in the configured target environment ("chrome87", "edge88", "es2020", "firefox78", "safari14" + 2 overrides)
> 16.83 19237|    locale: "en",
> 16.83 19238|    fallbackLocale: "en",
> 16.83 19239|    messages: await loadLocaleMessages()
> 16.83    |              ^
> 16.83 19240|  });
> 16.83 19241|
> 16.83
> 16.83     at failureErrorWithLog (/app/node_modules/esbuild/lib/main.js:1472:15)
> 16.83     at /app/node_modules/esbuild/lib/main.js:755:50
> 16.83     at responseCallbacks.<computed> (/app/node_modules/esbuild/lib/main.js:622:9)
> 16.83     at handleIncomingPacket (/app/node_modules/esbuild/lib/main.js:677:12)
> 16.83     at Socket.readFromStdout (/app/node_modules/esbuild/lib/main.js:600:7)
> 16.83     at Socket.emit (node:events:513:28)
> 16.83     at addChunk (node:internal/streams/readable:315:12)
> 16.83     at readableAddChunk (node:internal/streams/readable:289:9)
> 16.83     at Socket.Readable.push (node:internal/streams/readable:228:10)
> ```
> 
> again, not sure if I'm doing things right (following readme instructions).
> 
> Also a ton of SASS related deprecation warnings get thrown but that's probably something to sort with the bulk code changes with library updates later.
> 
> @amanji cleaned some stuff up in Traction related to this in the past (I think, we did remove that loadLocaleMessages function to use the plugin), can see what he did here if it applies: [bcgov/traction#630](https://github.com/bcgov/traction/pull/630)

@Gavinok I'd suggest updating the `vite.config.ts` according to: https://stackoverflow.com/questions/72618944/get-error-to-build-my-project-in-vite-top-level-await-is-not-available-in-the

Can test the application build with `npm run build` to ensure it works

amanji commented:
If we can resolve the build issue, otherwise everything LGTM

loneil commented:
Oh one other thing to consider, since changing the build process is the dockerfile that runs build (at least I assume that's the build we're dealing with) is on `ARG build_image=node:gallium` which is Node 16.
So could switch to 20 or 22 for the build if we're updating the build to use Vite anyways. (or at least track an issue to update, but then would have to retest)

Also note devcontainer (I didn't try that, not sure if it's in use) has 16 as well.
And looks like it adds vetur, where Volar should be used for Vue3, so can consider that for an update when switching the Vue version.

amanji commented:
Agreed with above.

loneil commented:
The babel and jest config files in the project still have `vue/cli-plugin` references, those configs might not be needed using Vite to build?

Gavinok committed:
Updated docker file version and vite build support

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

Gavinok commented:
I have made the requested chances to the docker file, updated the vite config, and removed the unused files. Hopefully that covers everything remaining. I haven't managed to get von running but if @amanji is able to confirm that it starts up and runs as expected this should be ready to merge

loneil commented on package.json:
```diff
@@ -54,11 +48,9 @@
-    "vue-template-compiler": "^2.7.16",
+    "typescript": "~5.6.3",
+    "unplugin-vue-components": "^0.27.4",
+    "vite": "^5.4.11",
```
Not sure if worth swapping on this PR but Vite 6 has since been released
https://vite.dev/blog/announcing-vite6

Gavinok replied:
the vue 2 plugin does not support it yet

loneil replied:
👍

amanji commented:
A couple of things:

1. The docker build is failing:

![image](https://github.com/user-attachments/assets/76f2882c-f66c-4019-9371-d1f323d5ab51)

2. `npm run build` will successfully complete, however the served bundle (using `npx vite preview`) shows an error resolving `require`

<img width="2516" alt="image" src="https://github.com/user-attachments/assets/8cd9ce9b-a05b-4d7d-9589-d992bb82bf4e">

Running in development seems to work fine.

Gavinok committed:
Fix image asset importing

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

Gavinok committed:
Fixed failed preview due to bug in vite vue2 plugin

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

Gavinok committed:
add explanation to the issue in the vue2 plugin

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

Gavinok commented:
@amanji Seems like an issue with the vue2 plugin. I incorporated a workaround until we can move to vue3

esune reviewed (approved):
Did a light testing, application builds correctly and starts up as expected.

esune commented:
> Did a light testing, application builds correctly and starts up as expected.

Actually having build issues on a different machine, so checking further

esune reviewed (changes requested):
Confirming there still seems to be an issue with the build:
```
2.947 x Build failed in 1.64s
2.947 error during build:
2.947 [vite]: Rollup failed to resolve import "graphic" from "/app/src/components/search/SearchHome.vue".
2.947 This is most likely unintended because it can break your application at runtime.
2.947 If you do want to externalize this module explicitly add it to
2.947 `build.rollupOptions.external`
2.947     at viteWarn (file:///app/node_modules/vite/dist/node/chunks/dep-CB_7IfJ-.js:65589:17)
2.947     at onRollupWarning (file:///app/node_modules/vite/dist/node/chunks/dep-CB_7IfJ-.js:65621:5)
2.947     at onwarn (file:///app/node_modules/vite/dist/node/chunks/dep-CB_7IfJ-.js:65284:7)
2.947     at file:///app/node_modules/rollup/dist/es/shared/node-entry.js:19472:13
2.947     at Object.logger [as onLog] (file:///app/node_modules/rollup/dist/es/shared/node-entry.js:21198:9)
2.947     at ModuleLoader.handleInvalidResolvedId (file:///app/node_modules/rollup/dist/es/shared/node-entry.js:20087:26)
2.947     at file:///app/node_modules/rollup/dist/es/shared/node-entry.js:20045:26
```

Gavinok commented:
I could have sworn I just fixed that. I'll give it another try. There is an unresolved bug in the vue 2 plugin I've been working around

Gavinok committed:
Corrected graphics reference

Signed-off-by: Gavin Jaeger-Freeborn <gavinfreeborn@gmail.com>

esune committed:
Fix typo in README

Signed-off-by: Emiliano Suñé <emiliano.sune@gmail.com>

esune reviewed (approved):
Tested locally, builds and runs correctly.
